lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,516 @@
+
+  if (isUnicodeRegionSubtag(getCurrentSubtagRef())) {
+    languageId.regionSubtag = getCurrentSubtag();
+    if (!transformedExtensionId) {
+      toASCIIUpperCase(languageId.regionSubtag);
+    }
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+  
+  while (true) {
+    if (!isUnicodeVariantSubtag(getCurrentSubtagRef())) {
+      return true;
+    } else if (!addVariantSubtag(transformedExtensionId)) {
+      return false;
+    }
+  
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+}
+
+bool LanguageTagParser::addVariantSubtag(bool transformedExtensionId) {
+  ParsedLanguageIdentifier &languageId = transformedExtensionId ? parsedLocaleIdentifier.transformedLanguageIdentifier : parsedLocaleIdentifier.languageIdentifier;
+
+  // Insert in alphabetical order
+  if (languageId.variantSubtagList.empty()) {
+    languageId.variantSubtagList.push_back(getCurrentSubtag());
+  } else {
+    auto subtag = getCurrentSubtag();
+    auto begin = languageId.variantSubtagList.begin();
+    auto end = languageId.variantSubtagList.end();
+    auto position = std::upper_bound(begin, end, subtag);
+    if (position != std::lower_bound(begin, end, subtag)) {
+      return false;
+    }
+    languageId.variantSubtagList.insert(position, subtag);
+  }
+  return true;
+}
+
+bool LanguageTagParser::parseExtensions() {
+  while (true) {
+    if (subtagEnd_ != subtagStart_) {
+      return true;
+    }
+    
+    char16_t singleton = getCurrentSubtag()[0];
+    if (!isASCIILetterOrDigit(singleton)) {
+      return true;
+    }
+    
+    if (!nextSubtag()) {
+      return false;
+    }
+    
+    // duplicate subtags handled in parse functions
+    switch (singleton) {
+      case 'u': {
+        // unicode extension
+        if (!parseUnicodeExtension()) {
+          return false;
+        }
+        break;
+      }
+      case 't': {
+        // transformed extension
+        if (!parseTransformedExtension()) {
+          return false;
+        }
+        break;
+      }
+      case 'x': {
+        // private use extension
+        if (!parsePUExtension()) {
+          return false;
+        }
+        return true;
+      }
+      default: {
+        // other extension
+        if (!parseOtherExtension(singleton)) {
+          return false;
+        }
+        break;
+      }
+    }
+  }
+
+  return false;
+}
+
+// unicode_locale_extensions = sep [uU]
+// ((sep keyword)+
+// |(sep attribute)+ (sep keyword)*) ;
+//
+// keyword = = key (sep type)? ;
+//
+// key = = alphanum alpha ;
+//
+// type = = alphanum{3,8}
+//  (sep alphanum{3,8})* ;
+//
+// attribute = alphanum{3,8} ;
+bool LanguageTagParser::parseUnicodeExtension() {
+  if (!parsedLocaleIdentifier.unicodeExtensionAttributes.empty() ||
+      !parsedLocaleIdentifier.unicodeExtensionKeywords.empty()) {
+    return false;
+  }
+  
+  while (isUnicodeExtensionAttribute(getCurrentSubtagRef())) {
+    // Insert in sorted order
+    if (parsedLocaleIdentifier.unicodeExtensionAttributes.empty()) {
```

Comment:
Similarly here, it's probably better to sort this at the end, or once it is needed. If sorting is explicitly only needed for canonicalisation, then we can also do it there.